lib/src/main/kotlin/com/swmansion/starknet/data/types/transactions/TransactionReceipt.kt

```diff
@@ -70,29 +71,38 @@ sealed class TransactionReceipt {
     abstract val events: List<Event>
     abstract val messagesSent: List<MessageL2ToL1>
     abstract val executionResources: ExecutionResources
+    abstract val blockHash: Felt?
+    abstract val blockNumber: Int?
 
     val isAccepted: Boolean
         get() = (
             executionStatus == TransactionExecutionStatus.SUCCEEDED &&
                 (finalityStatus == TransactionFinalityStatus.ACCEPTED_ON_L1 || finalityStatus == TransactionFinalityStatus.ACCEPTED_ON_L2)
             )
-    abstract val isPending: Boolean
-}
 
-@Serializable
-sealed class ProcessedTransactionReceipt : TransactionReceipt() {
-    abstract val blockHash: Felt
-    abstract val blockNumber: Int
-    override val isPending: Boolean = false
-}
-
-@Serializable
-sealed class PendingTransactionReceipt : TransactionReceipt() {
-    override val isPending: Boolean = true
+    /**
+     * Checks if the receipt has block info.
+     *
+     * Checks if the receipt matches `TXN_RECEIPT_WITH_BLOCK_INFO` schema from the JSON-RPC spec.
+     **
+     * @return true if both [blockHash] and [blockNumber] are not null, false otherwise.
+     */
+    val hasBlockInfo get() = listOf(blockHash, blockNumber).all { it != null }
+
+    /**
+     * Checks if the transaction is pending.
+     *
+     * Returns correct result only for receipts obtained using [Provider.getTransactionReceipt].
+     * For receipts obtained using [Provider.getBlockWithReceipts], use [BlockWithReceipts.isPending].
+     *
+     * @return true if the transaction is pending, false otherwise.
+     */
+    val isPending: Boolean
+        get() = !hasBlockInfo && finalityStatus == TransactionFinalityStatus.ACCEPTED_ON_L2
```

Comment:
Considering Argent is interacting with receipts from top-level, and other users probably do to, I assumed we shouldn't fully eliminate `isPending` property. 

It, however, is problematic in the 0.7 spec, as the receipt in `ProcessedBlockWithReceipts` (`BLOCK_BODY_WITH_RECEIPTS`) can satisfy all the conditions.

My suggestions are:
1. Add `blockHash` and `blockNumber` from `ProcessedBlockWithReceipts` to each of the receipts. That will however involve costly `copy` operations. It can be checkout at a471b657eac9c516bb831c621dd098769d232b03.
2. Current approach:
  - Add `Block.isPending` method
  - Add info to the docs that `TransactionReceipt.isPending` shouldn't really be used when the receipt is obtained from `BlockWithReceipts.isPending`

Reply:
I've already approved, but on second thoughts, the current approach is indeed not ideal and we should think about something better. Do we have any rough estimates of the impact of the first approach?

Reply:
There are also discussions to add `block_number` to pending receipt. Then, we could potentially make `isPending` throw error or return `null` if it's obtained using `Provider.getBlockWithReceipts`. I think this is all however a topic for a separate discussion when the decision is made and the new version of the spec is released. As for now I suggest we pick one of the currently available solutions to not further delay the release

> Do we have any rough estimates of the impact of the first approach?

I further read about `.copy()` and it seems like it's not as costly as I thought. Do you think we should go for it instead?

Reply:
If there's no premise to suggest a significant performance impact, I believe we should go with the first approach for now. Though it's a bit of a hack, it assures a consistent user experience, ensuring identical behavior regardless of the method used to get a transaction receipt.